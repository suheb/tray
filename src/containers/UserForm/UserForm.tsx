import React, { useState, ChangeEvent } from "react";
import { TextField, Grid, Button } from "@material-ui/core";

interface UserFormProps {
  formData?: { [field: string]: any }
  handleSubmit(formName:string, formData: { [field: string]: any }): void
}
interface InputConfig {
  regex: RegExp,
  error: string
}

const UserForm = (props: UserFormProps) => {
  const formInputConfig: {[field: string]: InputConfig} = {
    name: {
      regex: /\S/,
      error: "Required"
    },
    email: {
      regex: /\S+@\S+\.\S+/,
      error: "Invalid email"
    },
    password: {
      regex: /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])[\S]{9,}$/,
      error: "Must have atleast 9 characters with atleast 1 number, 1 uppercase + lowercase character."
    }
  }
  const intialFormData = {
    name: "",
    role: "",
    email: "",
    password:""
  }

  const [formData, setFormData] = useState<{ [field: string]: any }>(props.formData || intialFormData);
  const [formError, setFormError] = useState<{ [field: string]: string }>({});

  const isInputValid = (name: string, value: string) => {
    if (name in formInputConfig) {
      return formInputConfig[name].regex.test(value)
    }
    return true
  }

  const handleChange = (event: ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value
    const name = event.target.id
    // Set error
    if (isInputValid(name, value)) {
      const newFormError = {...formError};
      delete newFormError[name]
      setFormError(newFormError)
    } else {
      setFormError({
        ...formError,
        [name]: formInputConfig[name] ? formInputConfig[name].error : "Invalid"
      })
    }

    // Set field value
    setFormData({
      ...formData,
      [name]: value
    })
  }
  const handleSubmit = () => {
    const newFormError = { ...formError }
    Object.keys(intialFormData).forEach(fieldName => {
      if (!isInputValid(fieldName, formData[fieldName])) {
        newFormError[fieldName] = formInputConfig[fieldName] ? formInputConfig[fieldName].error : "Invalid"
      }
    })
    if (Object.keys(newFormError).length === 0) {
      props.handleSubmit("user", formData)
    } else {
      setFormError(newFormError)
    }
  }
  return (
    <Grid
      container
      justify="flex-end"
      spacing={2}
    >
      <Grid
        item
        xs={12}
      >
        <TextField
          required
          fullWidth
          id="name"
          label="Name"
          value={formData["name"] || ""}
          error={"name" in formError}
          helperText={formError["name"]}
          onChange={handleChange}
        />
      </Grid>

      <Grid
        item
        xs={12}
      >
        <TextField
          fullWidth
          id="role"
          label="Role"
          value={formData["role"] || ""}
          error={"role" in formError}
          helperText={formError["role"]}
          onChange={handleChange}
        />
      </Grid>
      <Grid
        item
        xs={12}
      >
        <TextField
          fullWidth
          required
          id="email"
          label="Email"
          type="email"
          value={formData["email"] || ""}
          error={"email" in formError}
          helperText={formError["email"]}
          onChange={handleChange}
        />
      </Grid>
      <Grid
        item
        xs={12}
      >
        <TextField
          fullWidth
          required
          id="password"
          label="Password"
          type="password"
          value={formData["password"] || ""}
          error={"password" in formError}
          helperText={formError["password"]}
          onChange={handleChange}
        />
      </Grid>
      <Grid
        item
        xs={3}
      >
        <Button
          fullWidth
          variant="contained"
          color="primary"
          onClick={handleSubmit}
        >
          Submit
        </Button>
      </Grid>
    </Grid>
  )
}

export default UserForm;
