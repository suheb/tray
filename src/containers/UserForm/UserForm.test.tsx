import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event'

import UserForm from './UserForm';

describe("UserForm", () => {
  it("renders without crashing", () => {
    const handleSubmit = jest.fn();
    const { asFragment } = render(<UserForm handleSubmit={handleSubmit} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("does not submit if there are errors", () => {
    const handleSubmit = jest.fn();
    const { getByText } = render(<UserForm handleSubmit={handleSubmit} />);
    userEvent.click(getByText(/submit/i));
    expect(handleSubmit).not.toBeCalled()
  });

  it("does submits with correct values", async () => {
    const handleSubmit = jest.fn();
    const formData = {
      name: 'John Smith',
      role: 'Analyst',
      email: 'abc@gmail.com',
      password: '1Abcdefdag'
    }
    const { getByText, getByLabelText } = render(<UserForm handleSubmit={handleSubmit} />);
    await userEvent.type(getByLabelText(/name/i), "John Smith")
    await userEvent.type(getByLabelText(/role/i), "Analyst")
    await userEvent.type(getByLabelText(/email/i), "abc@gmail.com")
    await userEvent.type(getByLabelText(/password/i), "1Abcdefdag")

    userEvent.click(getByText(/submit/i));
    expect(handleSubmit).toBeCalledWith("user", formData)
  });
});
