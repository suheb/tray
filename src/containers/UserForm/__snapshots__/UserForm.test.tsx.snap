// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-justify-xs-flex-end"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
          data-shrink="false"
          for="name"
          id="name-label"
        >
          Name
          <span
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="name"
            required=""
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="role"
          id="role-label"
        >
          Role
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="role"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
          data-shrink="false"
          for="email"
          id="email-label"
        >
          Email
          <span
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="email"
            required=""
            type="email"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
          data-shrink="false"
          for="password"
          id="password-label"
        >
          Password
          <span
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="password"
            required=""
            type="password"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Submit
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
